Rewrote components for React Context API v16.3+

ijxy opened the pull request:
The existing components are written for the legacy React Context API
according to the drizzle-react documentation.

The components provided here follow the v16.3+ standard, i.e. passing
`drizzle` and `drizzleState` as props. These variables are managed by a
DrizzleContext class that wraps around the body of an
application. Please see drizzle-react for more information.

I have also fixed these bugs in the current components:
* displaying array/objects in ContractData
* handling bytes32 data in ContractForm

I have added a `displayFunc` prop to ContractData and AccountData to
enable custom output formatting.

I have removed LoadingContainer as it is no longer necessary in v16.3+
(use DrizzleContext.Provider and DrizzleContext.Consumer instead).

ijxy committed:
Rewrote components for React Context API v16.3+

The existing components are written for the legacy React Context API
according to the drizzle-react documentation.

The components provided here follow the v16.3+ standard, i.e. passing
`drizzle` and `drizzleState` as props. These variables are managed by a
DrizzleContext class that wraps around the body of an
application. Please see drizzle-react for more information.

I have also fixed these bugs in the current components:
* displaying array/objects in ContractData
* handling bytes32 data in ContractForm

I have added a `displayFunc` prop to ContractData and AccountData to
enable custom output formatting.

I have removed LoadingContainer as it is no longer necessary in v16.3+
(use DrizzleContext.Provider and DrizzleContext.Consumer instead).

adrianmcli commented:
Thanks @parkinsonj!!! This looks like quite a bit of work, it might take me a couple days to get through this. Thanks for doing this though.

parkinsonj commented:
@adrianmcli You're welcome! I've spotted a couple of typos in the README, in the example usage parts: it should always say `drizzleState` but I've put `drizzleStatus` a few times. Let me know if you want me to push a fix.

adrianmcli commented:
@parkinsonj yeah go ahead and keep pushing whatever fixes you find

ijxy committed:
Update README.md

adrianmcli commented on README.md:
````diff
@@ -1,36 +1,98 @@
+
+### AccountData
+
+```<AccountData drizzle={drizzle} drizzleState={drizzleState} accountIndex='0' units='ether' displayFunc={(addr, bal, units) => (<div><strong>{addr}</strong>: {bal} {units}</div>)} />```
````
In this example, when you put `ether` for the `units` prop, it doesn't actually change the output balance. It simply changes the string from `Wei` to `Ether`. Might be better to call this string `displayUnits` instead.

ijxy replied:
This has been fixed together with the precision. I'd accidentally taken out the logic that called `web3.utils.fromWei(balance, units)` to actually change the values. Added it back now.

adrianmcli commented on README.md:
````diff
@@ -1,36 +1,98 @@
+<ul><li>structKey1</li><li>structKey2</li><li>...</li></ul>
+
+```
+<ContractData drizzle={drizzle} drizzleState={drizzleState} contract='MyContract' method='getStruct' methodArgs=['1'] />
````
The `methodArgs` prop is invalid syntax. It should be `methodArgs={['1']}` instead of `methodArgs=['1']`

ijxy replied:
Fixed.

adrianmcli commented on src/AccountData.js:
```diff
@@ -1,68 +1,42 @@
-    }
-
-    // Adjust to given precision.
-    if (this.props.precision) {
```
We're lacking the `precision` prop from the old implementation. This should fix the wrong decimals issue.

ijxy replied:
Fixed this one.

adrianmcli reviewed (changes requested):
I've only tried to replicate the `AccountData` component and the `ContractForm` and `ContractData` for the SimpleStorage contract so far. I might ask for more changes again, but if you have the time perhaps fix these first. Thanks again for your hard work!

ijxy committed:
Fixed Precision and Units

Restored logic for adjusting precision and converting the balance to different units.

ijxy committed:
Update README.md

Fixed syntax in example.

parkinsonj commented:
@adrianmcli Thanks for finding those - I've pushed some fixes. Keep shooting requests across if/when you find anything else!

adrianmcli commented on src/AccountData.js:
```diff
@@ -21,14 +17,24 @@ class AccountData extends Component {
 	render() {
+		// Use lowercase units. See https://web3js.readthedocs.io/en/1.0/web3-utils.html#fromwei.
+		const units = this.props.units ? this.props.units.toLowerCase() : 'wei'
+		var balance = this.context.drizzle.web3.utils.fromWei(this.state.balance, units)
```
`this.context` should be `this.props`

ijxy replied:
Turns out there was another issue hidden here. The default value of `null` for the balance was throwing an error in web3.utils.fromWei; I've gone ahead and changed the default to 0, which fixes this.

adrianmcli commented:
@parkinsonj It looks like something broke with the latest commits (see comments above). I think it would be a good idea for you to test to make sure all these new components work before we merge this PR.

I've created a simple project for testing here: https://github.com/adrianmcli/simple-drizzle-box

You'll notice that it currently implements the AccountData component, and the ContractForm and ContractData components for the SimpleStorage contract. I've invited you as a collaborator, let me know if you have any questions. The goal is to make sure that your new components are working properly in a drizzle app before we merge it.

ijxy committed:
Fixed bug in converting balance to different units

parkinsonj commented:
@adrianmcli  I've fixed the bug and tested against the dummy repo you set up. Screenshot below. Pushing the changes now. (If you run into issues with drizzle-react-components when running the server, cd into `client/node_modules/drizzle-react-components` then run `npm install` followed by `npm run-script build`.)

![image](https://user-images.githubusercontent.com/32483798/48804879-b7def480-ed0d-11e8-8a99-ab2f395f95e3.png)

adrianmcli commented:
@parkinsonj thanks! I'll test this one again soon.

adrianmcli commented:
This looks functional. I'm going to merge this into the `next` branch. And once everything is nice and tidy, we'll publish a new version. Thanks @parkinsonj!

parkinsonj commented:
@adrianmcli excellent! I hope to be able to contribute again soon :smile:

carlosescom commented:
Awesome contribuiton @parkinsonj, I had been trying to solve an issue with ContractData where as it didn't update it's DataKey after constructor invocation. Then I stumbled upon this magnificient contribution. And I saw you followed React Context API v16.3+ spec, so I knew I definitely needed to rewrite my code on top you what you did.

I hope this is not a bad place to comment this @adrianmcli, but I just wanted to give proper recognition to such important contributions and invite you both to review my PR.
